File: src/controllermenu.py
from typing import TYPE_CHECKING
from .menu import Menu
from .buttons import TextButton
from .selector import Selector
from .common import GameState
import pygame

if TYPE_CHECKING:
    from .game import Game
    from .controls import Controls

class JoystickEntry:
    def __init__(self):
        self.layout = None 
        self.define = None 

class ControllerMenu(Menu):
    def __init__(self, game: 'Game'):
        super().__init__(game)
        
        self._controls = game.get_controls()
        
        self._keyboard = [None, None]
        self._keyboard[0] = TextButton(self, 2.0, 4.6, 0.5, "Edit Layout")
        self._keyboard[1] = TextButton(self, 2.0, 3.8, 0.5, "Edit Layout")
        
        total_ctrls = 2 + pygame.joystick.get_count()
        self._joysticks = []
        
        for i in range(total_ctrls - 2):
            entry = JoystickEntry()
            entry.layout = Selector(self, 0.0, 1.1 - (i * 0.8), 3.3, 0.5)
            entry.define = TextButton(self, 4.5, 1.1 - (i * 0.8), 0.5, "Edit Layout")
            
            for l in range(1, 9):
                entry.layout.add_option(f"Layout {l}")
                
            current_layout = self._controls.get_layout(i + 2)
            entry.layout.set_option(current_layout - 2)
            
            self._joysticks.append(entry)
            
        self._back_button = TextButton(self, 0.0, -6.0, 0.7, "Back")

    def update(self, time: float) -> int:
        self.update_background(time)
        
        for i in range(2):
            if self._keyboard[i].update():
                self._game.set_active_controller(i)
                return GameState.SET_CONTROLS_MENU
                
        for i in range(len(self._joysticks)):
            self._joysticks[i].layout.update()
            
            if self._joysticks[i].define.update():
                layout = self._joysticks[i].layout.get_option()
                self._game.set_active_controller(layout + 2)
                return GameState.SET_CONTROLS_MENU
                
        if self._back_button.update():
            for i in range(len(self._joysticks)):
                self._controls.set_layout(i + 2, self._joysticks[i].layout.get_option() + 2)
                
            self._game.get_controls_file().write_file()
            return GameState.OPTION_MENU
            
        return GameState.CURRENT_STATE

    def draw(self):
        self.draw_background()
        dark = (0, 0, 0, 128)
        brown = (153, 76, 0, 128)
        
        boxes = [
            [(-7.0, 3.2), (7.0, 3.2), (7.0, 6.0), (-7.0, 6.0)],
            [(-7.0, -5.2), (7.0, -5.2), (7.0, 2.8), (-7.0, 2.8)],
            [(-7.0, -6.6), (7.0, -6.6), (7.0, -5.4), (-7.0, -5.4)]
        ]
        
        for b in boxes:
            self.draw_game_polygon(b, dark)
            
        brown_boxes = [
            [(-4.0, -6.4), (4.0, -6.4), (4.0, -5.6), (-4.0, -5.6)],
            [(-6.4, 3.4), (6.4, 3.4), (6.4, 4.1), (-6.4, 4.1)],
            [(-6.4, 4.2), (6.4, 4.2), (6.4, 4.9), (-6.4, 4.9)]
        ]
        
        for b in brown_boxes:
            self.draw_game_polygon(b, brown)
            
        for i in range(8):
             y_base = 0.75 - i * 0.8
             pts = [(-6.4, y_base), (6.4, y_base), (6.4, y_base + 0.7), (-6.4, y_base + 0.7)]
             self.draw_game_polygon(pts, brown)

        self._ui.draw_centered_text(0.0, 6.5, "Set Controls", style=self._ui.style(0.6, (1.0, 1.0, 1.0), shadow=True))
        info_style = self._ui.style(0.4, (0.5, 0.5, 0.5), spacing=0.3)
        self._ui.draw_centered_text(0.0, 2.2, "Joystick", style=info_style)
        self._ui.draw_centered_text(0.0, 1.8, "layout number", style=info_style)
        self._ui.draw_centered_text(4.5, 2.2, "Change", style=info_style)
        self._ui.draw_centered_text(4.5, 1.8, "joystick layout", style=info_style)
        self._ui.draw_centered_text(2.0, 5.4, "Change", style=info_style)
        self._ui.draw_centered_text(2.0, 5.0, "keyboard layout", style=info_style)

        keyboard_style = self._ui.style(0.6, (0.0, 1.0, 1.0))
        self._ui.draw_centered_text(-2.0, 4.3, "Keyboard 1", style=keyboard_style)
        self._ui.draw_centered_text(-2.0, 3.5, "Keyboard 2", style=keyboard_style)

        disconnected_style = self._ui.style(0.5, (0.5, 0.5, 0.5), spacing=0.4)
        connected_joys = len(self._joysticks)
        for i in range(connected_joys, 8):
            self._ui.draw_centered_text(-4.6, 0.8 - i*0.8, f"Joystick {i+1}", style=disconnected_style)
            self._ui.draw_centered_text(2.0, 0.8 - i*0.8, "<<Not Connected>>", style=disconnected_style)
            
        for i in range(connected_joys):
            self._ui.draw_centered_text(-4.6, 0.8 - i*0.8, f"Joystick {i+1}", style=keyboard_style)
            self._joysticks[i].layout.draw()
            self._joysticks[i].define.draw()
            
        self._keyboard[0].draw()
        self._keyboard[1].draw()
        self._back_button.draw()

File: groundfire-0.25/src/controllermenu.cc
////////////////////////////////////////////////////////////////////////////////
//
//               Groundfire
//
////////////////////////////////////////////////////////////////////////////////
//
// Copyright (c) 2004, Tom Russell (tom@groundfire.net)
//
// This file is part of the Groundfire project, distributed under the MIT 
// license. See the file 'COPYING', included with this distribution, for a copy
// of the full MIT licence.
//
////////////////////////////////////////////////////////////////////////////////
//
//   File name : controllermenu.cc
//
//          By : Tom Russell
//
//        Date : 09-Nov-03
//
// Description : Handles the controller configuration menu accessible from the
//               options menu.
//
//
////////////////////////////////////////////////////////////////////////////////

////////////////////////////////////////////////////////////////////////////////
// Includes
////////////////////////////////////////////////////////////////////////////////
#include "controllermenu.hh"
#include "font.hh"

////////////////////////////////////////////////////////////////////////////////
// Public Member Functions
////////////////////////////////////////////////////////////////////////////////

////////////////////////////////////////////////////////////////////////////////
//
// Function    : cControllerMenu
//
// Description : Constructor
//
////////////////////////////////////////////////////////////////////////////////
cControllerMenu::cControllerMenu 
(
    cGame * game
)
: cMenu (game)
{

    // Grab a pointer to the controls object because we'll be using this a
    // lot.
    _controls = game->getControls ();

    // Add the buttons for redefining the keyboard.
    _keyboard[0] = new cTextButton (this, 2.0f, 4.6f, 0.5f, "Edit Layout");
    _keyboard[1] = new cTextButton (this, 2.0f, 3.8f, 0.5f, "Edit Layout");

    // Now add the buttons for the joysticks
    for (int i = 0; i < _interface->numOfControllers () - 2; i++) 
    {
        // Create a selector for the different joystick layout options
        _joystick[i].layout = new cSelector (this, 0.0f, 1.1f - (i * 0.8f),
                                               3.3f, 0.5f);
        
        // Create a button for Editing the joystick layout
        _joystick[i].define = new cTextButton (this, 
                                               4.5f, 1.1f - (i * 0.8f), 0.5f,
                                               "Edit Layout");

        // Add the layout options for the joysticks
        _joystick[i].layout->addOption ("Layout 1");
        _joystick[i].layout->addOption ("Layout 2");
        _joystick[i].layout->addOption ("Layout 3");
        _joystick[i].layout->addOption ("Layout 4");
        _joystick[i].layout->addOption ("Layout 5");
        _joystick[i].layout->addOption ("Layout 6");
        _joystick[i].layout->addOption ("Layout 7");
        _joystick[i].layout->addOption ("Layout 8");

        // Set the layout selector to the current layout for this joystick.
        _joystick[i].layout->setOption (_controls->getLayout (i + 2) - 2);
    }

    _backButton = new cTextButton (this, 0.0f, -6.0f, 0.7f, "Back");
}

////////////////////////////////////////////////////////////////////////////////
//
// Function    : ~cControllerMenu
//
// Description : Destructor
//
////////////////////////////////////////////////////////////////////////////////
cControllerMenu::~cControllerMenu 
(
)
{
    delete _backButton;

    delete _keyboard[0];
    delete _keyboard[1];

    for (int i = 0; i < _interface->numOfControllers () - 2; i++)
    {
        _controls->setLayout (i + 2, _joystick[i].layout->getOption () + 2);

        delete _joystick[i].layout;
        delete _joystick[i].define;
    }
}

////////////////////////////////////////////////////////////////////////////////
//
// Function    : update
//
// Description : Update the menu. Returns the game state to change to (or 
//               CURRENT_STATE for no change.) 
//
////////////////////////////////////////////////////////////////////////////////
enumGameState
cControllerMenu::update
(
    double time // elapsed time since last update
)
{
    updateBackground (time);

    for (int i = 0; i < 2; i++)
    {
        // Check if either of the keyboard layout buttons have been clicked and
        // go to the set controls menu if they have.
        if (_keyboard[i]->update ())
        {
            // Tell the game object which controller we will be editing
            _game->setActiveController (i);

            return (SET_CONTROLS_MENU);
        }
    }
    
    for (int i = 0; i < _interface->numOfControllers () - 2; i++)
    {
        // Update the layout selector.
        _joystick[i].layout->update ();
        
        // Check if a joystick layout button has been clicked and go to the 
        // set controls menu if so.
        if (_joystick[i].define->update ())
        {
            // Which joystick layout are we editing?
            int layout = _joystick[i].layout->getOption ();
           
            // Tell the game object which controller we will be editing
            _game->setActiveController (layout + 2);

            return (SET_CONTROLS_MENU);
        }
    }

    if (_backButton->update ())
    {
        // Write out a new controls file with any of the changed layout
        // options.
        _game->getControlsFile ()->writeFile ();

        return (OPTION_MENU);
    }

    // No change of menu.
    return (CURRENT_STATE);
}

////////////////////////////////////////////////////////////////////////////////
//
// Function    : draw
//
// Description : Draw the menu
//
////////////////////////////////////////////////////////////////////////////////
void
cControllerMenu::draw
(
)

{
    drawBackground ();

    glEnable (GL_BLEND);
  
    // Draw the boxes
    
    glColor4f (0.0f, 0.0f, 0.0f, 0.5f);

    glBegin (GL_QUADS);

    glVertex3f   (-7.0f,  3.2f, 0.0f);
    glVertex3f   ( 7.0f,  3.2f, 0.0f);
    glVertex3f   ( 7.0f,  6.0f, 0.0f);
    glVertex3f   (-7.0f,  6.0f, 0.0f);

    glVertex3f   (-7.0f, -5.2f, 0.0f);
    glVertex3f   ( 7.0f, -5.2f, 0.0f);
    glVertex3f   ( 7.0f,  2.8f, 0.0f);
    glVertex3f   (-7.0f,  2.8f, 0.0f);

    glVertex3f   (-7.0f, -6.6f, 0.0f);
    glVertex3f   ( 7.0f, -6.6f, 0.0f);
    glVertex3f   ( 7.0f, -5.4f, 0.0f);
    glVertex3f   (-7.0f, -5.4f, 0.0f);

    glEnd ();

    // Draw the brown boxes under the buttons
    
    glColor4f (0.6f, 0.3f, 0.0f, 0.5f);

    glBegin (GL_QUADS);

    glVertex3f   (-4.0f, -6.4f, 0.0f);
    glVertex3f   ( 4.0f, -6.4f, 0.0f);
    glVertex3f   ( 4.0f, -5.6f, 0.0f);
    glVertex3f   (-4.0f, -5.6f, 0.0f);

    glVertex3f   (-6.4f, 3.4f, 0.0f);
    glVertex3f   ( 6.4f, 3.4f, 0.0f);
    glVertex3f   ( 6.4f, 4.1f, 0.0f);
    glVertex3f   (-6.4f, 4.1f, 0.0f);

    glVertex3f   (-6.4f, 4.2f, 0.0f);
    glVertex3f   ( 6.4f, 4.2f, 0.0f);
    glVertex3f   ( 6.4f, 4.9f, 0.0f);
    glVertex3f   (-6.4f, 4.9f, 0.0f);

    for (int i = 0; i < 8; i++) 
    {
        glVertex3f   (-6.4f, 0.75f - (i * 0.8f), 0.0f);
        glVertex3f   ( 6.4f, 0.75f - (i * 0.8f), 0.0f);
        glVertex3f   ( 6.4f, 1.45f - (i * 0.8f), 0.0f);
        glVertex3f   (-6.4f, 1.45f - (i * 0.8f), 0.0f);    
    }

    glEnd ();

    glDisable (GL_BLEND);

    // Draw the menu title
    
    _font->setSize (0.6f, 0.6f, 0.5f);
    _font->setColour (1.0f, 1.0f, 1.0f);
    _font->setShadow (true);

    _font->printCentredAt (0.0f, 6.5f, "Set Controls");
    
    _font->setShadow (false);
    _font->setSize (0.4f, 0.4f, 0.3f);
    _font->setColour (0.5f, 0.5f, 0.5f);

    // Write out some of the other text that appears in the menu
    _font->printCentredAt (0.0f, 2.2f, "Joystick");
    _font->printCentredAt (0.0f, 1.8f, "layout number");

    _font->printCentredAt (4.5f, 2.2f, "Change");
    _font->printCentredAt (4.5f, 1.8f, "joystick layout");

    _font->printCentredAt (2.0f, 5.4f, "Change");
    _font->printCentredAt (2.0f, 5.0f, "keyboard layout");

    _font->setSize (0.6f, 0.6f, 0.5f);
    _font->setColour (0.0f, 1.0f, 1.0f);
    _font->printCentredAt (-2.0f, 4.3f, "Keyboard 1");
    _font->printCentredAt (-2.0f, 3.5f, "Keyboard 2");

    // For joysticks that are not connected, write out a message in its place
    _font->setColour (0.5f, 0.5f, 0.5f);
    for (int i = _interface->numOfControllers () - 2; i < 8; i++) 
    {
        _font->printCentredAt (-4.6f, 0.8f - i * 0.8f, "Joystick %d", i + 1);
        _font->printCentredAt ( 2.0f, 0.8f - i * 0.8f, "<<Not Connected>>");
    }

    // Draw the keyboard layout buttons
    _keyboard[0]->draw ();
    _keyboard[1]->draw ();
    
    // Draw the connected joystick buttons/selectors
    for (int i = 0; i < _interface->numOfControllers () - 2; i++)
    {
        _font->setSize (0.6f, 0.6f, 0.5f);
        _font->setColour (0.0f, 1.0f, 1.0f);
        _font->printCentredAt (-4.6f, 0.8f - i * 0.8f, "Joystick %d", i + 1);
        
        _joystick[i].layout->draw ();
        _joystick[i].define->draw ();
    }

    _backButton->draw ();
}
